### PR TITLE
DT-643 Improve PNC matching and added CRO matching

### PR DESCRIPTION
**Pull request by andymarke**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/hmpps/offendersearch/services/MatchService.kt [hunk 1]
```diff
@@ -94,50 +114,66 @@ class MatchService(
...
-        .matchQuery(query, it))
+    this.must(matchQuery(query, it))
   }
   return this
 }
 
 fun BoolQueryBuilder.mustMultiMatchKeyword(value: String, vararg query: String): BoolQueryBuilder {
   this.must().add(multiMatchQuery(value, *query)
-      .analyzer("whitespace")
+      .analyzer("keyword")
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/hmpps/offendersearch/services/MatchService.kt]: decided to use "keyword" which is technically correct, the New Tech code uses "whitespace" which will achieve the same thing, but given the data is parsed as a keyword the ES documentation says we should use the same analyser as the parser

File: src/main/kotlin/uk/gov/justice/hmpps/offendersearch/services/MatchService.kt [hunk 2]
```diff
@@ -94,50 +114,66 @@ class MatchService(
...
   )
   return this
 }
 
+fun BoolQueryBuilder.mustKeyword(value: String, query: String): BoolQueryBuilder {
+  return this.must(matchQuery(query, value).analyzer("keyword"))
+}
 
-sealed class Result<out T, out M> {
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/hmpps/offendersearch/services/MatchService.kt]: Decided didn't need generics for this given it is specific to searching, makes the code a bit more readable

File: src/test/kotlin/uk/gov/justice/hmpps/offendersearch/controllers/OffenderMatchAPIIntegrationTest.kt [hunk 3]
```diff
@@ -58,7 +56,7 @@ internal class OffenderMatchControllerAPIIntegrationTest : AbstractTestExecution
   }
 
   @Nested
-  inner class `basic operation` {
+  inner class BasicOperation {
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/hmpps/offendersearch/controllers/OffenderMatchAPIIntegrationTest.kt]: In vain hope to a fix an issue where Intellij can't jump to specific test methods when there are many inner class tests - didn't work but this just as readble

File: src/test/kotlin/uk/gov/justice/hmpps/offendersearch/controllers/OffenderMatchAPIIntegrationTest.kt [hunk 4]
```diff
@@ -221,6 +219,199 @@ internal class OffenderMatchControllerAPIIntegrationTest : AbstractTestExecution
...
+          ))
+          .post("/match")
+          .then()
+          .statusCode(200)
+          .body("matches.findall.size()", equalTo(1))
+          .body("matches[0].offender.otherIds.crn", equalTo("X00001"))
+    }
+
+    @Test
+    internal fun `noms number takes precedence over CRO Number when present`() {
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/hmpps/offendersearch/controllers/OffenderMatchAPIIntegrationTest.kt]: How do you know that the nomsNumber took precedence?

File: src/main/kotlin/uk/gov/justice/hmpps/offendersearch/services/MatchService.kt [hunk 5]
```diff
@@ -29,25 +30,44 @@ class MatchService(
...
   private fun pncNumber(matchRequest: MatchRequest): BoolQueryBuilder? {
-    return matchRequest.pncNumber.takeIf { !it.isNullOrBlank() }?.let {
-      return QueryBuilders.boolQuery()
-          .mustMultiMatchKeyword(it.canonicalPNCNumber(), "otherIds.pncNumberLongYear", "otherIds.pncNumberShortYear")
+    with(matchRequest) {
+      return pncNumber.takeIf { !it.isNullOrBlank() }?.let {
+        return QueryBuilders.boolQuery()
+            .mustMultiMatchKeyword(it.canonicalPNCNumber(), "otherIds.pncNumberLongYear", "otherIds.pncNumberShortYear")
+            .must(QueryBuilders.boolQuery()
+                .shouldWhenPresent("surname", surname)
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/hmpps/offendersearch/services/MatchService.kt]: You said that should was coming but I don't understand the difference between must and should in a query context... Please can you elaborate just for my understanding :)

File: src/test/kotlin/uk/gov/justice/hmpps/offendersearch/controllers/OffenderMatchAPIIntegrationTest.kt (hunk 4 above)
Comment (andymarke) [reply src/test/kotlin/uk/gov/justice/hmpps/offendersearch/controllers/OffenderMatchAPIIntegrationTest.kt]: Very good question; so if we have a NOMS number it means a human has already matched this record. NOMS number only get into Delius by manually matching so this is really saying  "this record has already been linked" . Having said that I have definitely seen the occasional record that has been linked incorrectly - but for now decided not to tackle that particular edge case.

File: src/main/kotlin/uk/gov/justice/hmpps/offendersearch/services/MatchService.kt (hunk 5 above)
Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/hmpps/offendersearch/services/MatchService.kt]: Yep; in the Elastic Search language "must" = "and" while "should" = "or" therefore this statement says:

Match = "PNC Number" AND ("SURNAME" OR "DATE OF BIRTH")

In fact the JSON it creates looks something not too far from this.

By the way multi-match is similar; but means this "phrase" must match in one of these fields. So given the following
PNC = "2018/12345x"
SURNAME = "Marke"
DOB = "1965-07-19"

The expression above would be
```

WHERE (pncNumberLongYear = 2018/12345x OR pncNumberShortYear = 2018/12345x) 
AND (surname = MARKE OR dateOfBirth =  1965-07-19)

```

Comment (mikehalmamoj) [reply src/main/kotlin/uk/gov/justice/hmpps/offendersearch/services/MatchService.kt]: An interesting choice of semantics!  Thanks Andy.

**Pull request merged**